### DISCUSSION TITLE
Nginx proxy not working in some cases

### DISCUSSION THREAD

**Discussion by Daniel15** [plausible/analytics#3552] [Self-Hosted Support]
In order for Plausible log calls to be same-domain calls, I have an Nginx proxy configured like this:
```nginx
	# Plausible logging
	location = /api/event {
		proxy_pass https://hits.example.com/api/event;
		proxy_set_header Host hits.example.com;
		proxy_buffering on;
		proxy_http_version 1.1;

		proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
		proxy_set_header X-Forwarded-Proto $scheme;
		proxy_set_header X-Forwarded-Host $host;
		proxy_set_header Host hits.example.com;
		proxy_ssl_name hits.example.com;
		proxy_ssl_server_name on;
		proxy_ssl_session_reuse off;
	}
```

On one server, this is working fine. I see something like in the Nginx logs on my Plausible server:
```
2406:3a00:10::5bd7:36d5 - - [23/Nov/2023:16:01:21 -0800] "POST /api/event HTTP/1.1" 202 2 "https://mysite.com/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:120.0) Gecko/20100101 Firefox/120.0"
```

On another server, Nginx is not passing through the User-Agent header, resulting in a HTTP 400 error:
```
2a04:bdc7:100:95ed::1 - - [23/Nov/2023:16:01:54 -0800] "POST /api/event HTTP/1.1" 400 150 "-" "-"
```

The server hosting the actual site **does** receive the User-Agent:
```
2001:5a8:4170:xxxxxxx - - [23/Nov/2023:16:05:38 -0800] "POST /api/event HTTP/2.0" 400 150 "https://mysite2.net/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:120.0) Gecko/20100101 Firefox/120.0"
```
it's just not being passed by the reverse proxy to the upstream Plausible server for whatever reason.

I'm struggling to find the difference between the two server configurations. The sites are configured similarly, and the `/api/event` block is identical across both.

Any ideas? Has anyone seen this before?

**Comment by ruslandoga** [comment]
👋 @Daniel15 

Are there any differences in setup, i.e. nginx versions, how they are installed / configured, `nginx.conf`, etc.? One way to check where the header is dropped is to use curl from within the instance.

**Reply by ruslandoga** [reply]
> On another server, Nginx is not passing through the User-Agent header, resulting in a HTTP 400 error:

Does it happen every time or sometimes?

**Reply by ruslandoga** [reply]
> The server hosting the actual site does receive the User-Agent:

Do you have a two-tier system with a proxy in front of the two nginx instances? If so, the problem might be in the first proxy.

**Reply by Daniel15** [reply]
@ruslandoga  Sorry for the delay in replying! It turns out the server where it was working was running Debian 11 with Nginx 1.18, while the server where it wasn't working was running Debian 12 with Nginx 1.22.1. I upgraded the older server to Debian 12, and now they're both failing. So that's good (at least there's finally some consistency) but now I'm still not sure how to fix it.

I tried explicitly passing the User-Agent:
```
proxy_set_header User-Agent $http_user_agent;
```

but that didn't work - The Nginx logs on the Plausible server still show a blank user-agent. I tried a hard-coded string too (`proxy_set_header User-Agent HelloWorld`) which didn't work either.

I also tried `proxy_pass_request_headers on` which didn't work either.

I had a quick look at the [Git commits between 1.18 and 1.22.1](https://github.com/nginx/nginx/compare/release-1.18.0...release-1.22.1) but nothing really stuck out to me. Maybe some Debian-specific patch is causing it?

I'm stumped :/

One of the failing endpoints is here: https://api.dnstools.ws/api/event which is handled by this Nginx config: https://github.com/Daniel15/dnstools/blob/13df5eb35c1674e82e623bb8f383cc9b5d5e9a5c/src/DnsTools.Web/nginx/nginx.conf#L59

**Comment by Daniel15** [comment]
I worked around this issue by proxying directly to the Plausible backend rather than through the Nginx reverse proxy. That seems to work fine.

**Reply by darrenzou2000** [reply]
Hey @Daniel15 Can you write a small tutorial on how you did it?
